### PR TITLE
Apply P0740R0 PR for 261: Restrict alg.general changes from P0370 to …

### LINKED ISSUE #261: Restrict alg.general changes from P0370 to apply only to the range-and-a-half algorithms
..as directed during LWG review of P0370. LWG is uncomfortable with the uncertainty this wording introduces, and would prefer to "limit the scope of the potential damage" to only the deprecated range-and-a-half algorithms. 

Proposed resolution
--------------------
(This wording assumes the PRs of #286 and #379 have been applied)

Strike para [algorithms.general]/10 which begins: "Some algorithms declare both an overload that takes a Range and an Iterator, and an overload that takes two Range parameters."

Strike para [algorithms.general]/12 which begins, "Despite that the algorithm declarations nominally accept parameters by value [...]"

Replace the entire content of Annex A.2 [depr.algo.range-and-a-half] with:

> 1 The following algorithm signatures are deemed unsafe and are deprecated in this document.
>
> 2 Overloads of algorithms that take a `Range` argument and a forwarding reference parameter `first2_` behave as if they are implemented by calling `begin` and `end` on the `Range` and dispatching to the overload that takes separate iterator and sentinel arguments, perfectly forwarding `first2_`.
>
> ```c++
> template <InputIterator I1, Sentinel<I1> S1, class I2, class Pred = equal_to<>,
>     class Proj1 = identity, class Proj2 = identity>
>   requires InputIterator<decay_t<I2>> && !Range<I2> &&
>     IndirectPredicate<Pred, projected<I1, Proj1>, projected<decay_t<I2>, Proj2>>
>   tagged_pair<tag::in1(I1), tag::in2(decay_t<I2>)>
>     mismatch(I1 first1, S1 last1, I2&& first2_, Pred pred = Pred{},
>              Proj1 proj1 = Proj1{}, Proj2 proj2 = Proj2{});
>
> template <InputRange Rng1, class I2, class Pred = equal_to<>,
>     class Proj1 = identity, class Proj2 = identity>
>   requires InputIterator<decay_t<I2>> && !Range<I2> &&
>     IndirectPredicate<Pred, projected<iterator_t<Rng1>, Proj1>,
>         projected<decay_t<I2>, Proj2>>
>   tagged_pair<tag::in1(safe_iterator_t<Rng1>), tag::in2(decay_t<I2>)>
>     mismatch(Rng1&& rng1, I2&& first2_, Pred pred = Pred{},
>              Proj1 proj1 = Proj1{}, Proj2 proj2 = Proj2{});
> ```
>
> 3 *Effects:* Equivalent to: `return mismatch(first1, last1, std::forward<I2>(first2_), unreachable{}, pred, proj1, proj2);`, except that the underlying algorithm never increments `first2` more than `last1 - first1` times.
>
> ```c++
> template <InputIterator I1, Sentinel<I1> S1, class I2,
>     class Pred = equal_to<>, class Proj1 = identity, class Proj2 = identity>
>   requires InputIterator<decay_t<I2>> && !Range<I2> &&
>     IndirectlyComparable<I1, decay_t<I2>, Pred, Proj1, Proj2>
>   bool equal(I1 first1, S1 last1,
>              I2&& first2_, Pred pred = Pred{},
>              Proj1 proj1 = Proj1{}, Proj2 proj2 = Proj2{});
>
> template <InputRange Rng1, class I2, class Pred = equal_to<>,
>     class Proj1 = identity, class Proj2 = identity>
>   requires InputIterator<decay_t<I2>> && !Range<I2> &&
>     IndirectlyComparable<iterator_t<Rng1>, decay_t<I2>, Pred, Proj1, Proj2>
>   bool equal(Rng1&& rng1, I2&& first2_, Pred pred = Pred{},
>              Proj1 proj1 = Proj1{}, Proj2 proj2 = Proj2{});
> ```
>
> 4 *Effects:* Equivalent to: `return first1 == mismatch(first1, last1, std::forward<I2>(first2_), pred, proj1, proj2).in1();`
>
> ```c++
> template <ForwardIterator I1, Sentinel<I1> S1, class I2,
>     class Pred = equal_to<>, class Proj1 = identity, class Proj2 = identity>
>   requires ForwardIterator<decay_t<I2>> && !Range<I2> &&
>     IndirectlyComparable<I1, decay_t<I2>, Pred, Proj1, Proj2>
>   bool is_permutation(I1 first1, S1 last1, I2&& first2_, Pred pred = Pred{},
>                       Proj1 proj1 = Proj1{}, Proj2 proj2 = Proj2{});
>
> template <ForwardRange Rng1, class I2, class Pred = equal_to<>,
>     class Proj1 = identity, class Proj2 = identity>
>   requires ForwardIterator<decay_t<I2>> && !Range<I2> &&
>     IndirectlyComparable<iterator_t<Rng1>, decay_t<I2>, Pred, Proj1, Proj2>
>   bool is_permutation(Rng1&& rng1, I2&& first2_, Pred pred = Pred{},
>                       Proj1 proj1 = Proj1{}, Proj2 proj2 = Proj2{});
> ```
>
> 5 *Effects:* Equivalent to:
>
> ```c++
>   auto first2 = std::forward<I2>(first2_);
>   return is_permutation(first1, last1, first2, next(first2, distance(first1, last1)),
>                         pred, proj1, proj2);
> ```
>
> ```c++
> template <ForwardIterator I1, Sentinel<I1> S1, class I2>
>   requires ForwardIterator<decay_t<I2>> && !Range<I2> &&
>     IndirectlySwappable<I1, decay_t<I2>>
>   tagged_pair<tag::in1(I1), tag::in2(decay_t<I2>)>
>     swap_ranges(I1 first1, S1 last1, I2&& first2_);
>
> template <ForwardRange Rng, class I>
>   requires ForwardIterator<decay_t<I>> && !Range<I> &&
>     IndirectlySwappable<iterator_t<Rng>, decay_t<I>>
>   tagged_pair<tag::in1(safe_iterator_t<Rng>), tag::in2(decay_t<I>)>
>     swap_ranges(Rng&& rng1, I&& first2_);
> ```
>
> 6 *Effects:* Equivalent to:
>
> ```c++
>   auto first2 = std::forward<I2>(first2_);
>   return swap_ranges(first1, last1, first2, next(first2, distance(first1, last1)),
>                      pred, proj1, proj2);
> ```
>
> ```c++
> template <InputIterator I1, Sentinel<I1> S1, class I2, WeaklyIncrementable O,
>     class F, class Proj1 = identity, class Proj2 = identity>
>   requires InputIterator<decay_t<I2>> && !Range<I2> &&
>     Writable<O, indirect_result_of_t<F&(projected<I1, Proj1>,
>         projected<decay_t<I2>, Proj2>)>>
>   tagged_tuple<tag::in1(I1), tag::in2(decay_t<I2>), tag::out(O)>
>     transform(I1 first1, S1 last1, I2&& first2_, O result,
>               F binary_op, Proj1 proj1 = Proj1{}, Proj2 proj2 = Proj2{});
>
> template <InputRange Rng, class I, WeaklyIncrementable O, class F,
>     class Proj1 = identity, class Proj2 = identity>
>   requires InputIterator<decay_t<I>> && !Range<I> &&
>     Writable<O, indirect_result_of_t<F&(
>         projected<iterator_t<Rng>, Proj1>, projected<decay_t<I>, Proj2>>)>
>   tagged_tuple<tag::in1(safe_iterator_t<Rng>), tag::in2(decay_t<I>), tag::out(O)>
>     transform(Rng&& rng1, I&& first2_, O result,
>               F binary_op, Proj1 proj1 = Proj1{}, Proj2 proj2 = Proj2{});
> ```
>
> 7 *Effects:* Equivalent to: `return transform(first1, last1, std::forward<I2>(first2_), unreachable{}, pred, proj1, proj2);`, except that the underlying algorithm never increments `first2` more than `last1 - first1` times.

### PR DESCRIPTION
**Pull request by CaseyCarter**
…apply only to the range-and-a-half algorithms

Fixes #261.

### CODE REVIEW

File: deprecated.tex [hunk 1]
```diff
@@ -31,73 +31,124 @@
...
+\end{codeblock}
 
-template <ForwardIterator I1, Sentinel<I1> S1, ForwardIterator I2,
-    class Pred = equal_to<>, class Proj1 = identity, class Proj2 = identity>
-  requires IndirectlyComparable<I1, I2, Pred, Proj1, Proj2>
-  bool is_permutation(I1 first1, S1 last1, I2 first2,
-                      Pred pred = Pred{},
+\pnum
+\effects
+Equivalent to: \tcode{return first1 == mismatch(first1, last1, std::forward<I2>(first2_), comp, proj1, proj2).in1();}
```
Comment (ericniebler) [inline deprecated.tex]: Shouldn't this be `last1 == mismatch(....`?

Comment (CaseyCarter) [inline deprecated.tex]: Oh wow, colossal failure. Yes it should.

**Comment by CaseyCarter** [comment]
Merging this after fixing the bug found in review.

**Comment by ericniebler** [comment]
It's an obvious screw up, but this really isn't an editorial change. If you want to make this change, we need to keep a list of these changes and run them by the editorial committee. Handling this as a DR post-TS wouldn't be a big deal.

**Comment by CaseyCarter** [comment]
> It's an obvious screw up, but this really isn't an editorial change. If you want to make this change, we need to keep a list of these changes and run them by the editorial committee. Handling this as a DR post-TS wouldn't be a big deal.

Agreed. I've force-pushed master back a commit so this never happened, and recommitted without the bugfix.

**Pull request merged**